folding-schemes/src/folding/protogalaxy/circuits.rs

```diff
@@ -88,34 +107,363 @@ impl FoldingGadget {
+        C2: CurveGroup,
+        GC2: CurveVar<C2, CF2<C2>> + ToConstraintFieldGadget<CF2<C2>>,
+        S: CryptographicSponge,
+    >(
+        transcript: &mut PoseidonSpongeVar<CF1<C1>>,
+        pp_hash: FpVar<CF1<C1>>,
+        mut cf_U: CycleFoldCommittedInstanceVar<C2, GC2>,
+        cf_u_cmWs: Vec<GC2>,
+        cf_u_xs: Vec<Vec<NonNativeUintVar<CF1<C1>>>>,
+        cf_cmTs: Vec<GC2>,
+    ) -> Result<CycleFoldCommittedInstanceVar<C2, GC2>, SynthesisError>
+    where
+        C2::BaseField: PrimeField + Absorb,
+        for<'a> &'a GC2: GroupOpsBounds<'a, C2, GC2>,
+    {
+        assert_eq!(cf_u_cmWs.len(), cf_u_xs.len());
+        assert_eq!(cf_u_xs.len(), cf_cmTs.len());
+
+        // Fold the incoming CycleFold instances into the running CycleFold
+        // instance in a iterative way, since `NIFSFullGadget` only supports
+        // folding one incoming instance at a time.
+        for ((cmW, x), cmT) in cf_u_cmWs.into_iter().zip(cf_u_xs).zip(cf_cmTs) {
+            // Prepare the incoming CycleFold instance `cf_u` for the current
+            // iteration.
+            // For each CycleFold instance `cf_u`, we have `cf_u.cmE = 0`, and
+            // `cf_u.u = 1`.
+            let cf_u = CycleFoldCommittedInstanceVar {
+                cmE: GC2::zero(),
+                u: NonNativeUintVar::new_constant(ConstraintSystemRef::None, C1::BaseField::one())?,
+                cmW,
+                x,
+            };
+
+            let cf_r_bits = CycleFoldChallengeGadget::get_challenge_gadget(
+                transcript,
+                pp_hash.clone(),
+                cf_U.to_native_sponge_field_elements()?,
+                cf_u.clone(),
+                cmT.clone(),
+            )?;
+            // Fold the current incoming CycleFold instance `cf_u` into the
+            // running CycleFold instance `cf_U`.
+            cf_U = NIFSFullGadget::fold_committed_instance(cf_r_bits, cmT, cf_U, cf_u)?;
+        }
+
+        Ok(cf_U)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct AugmentedFCircuit<
+    C1: CurveGroup,
+    C2: CurveGroup,
+    GC2: CurveVar<C2, CF2<C2>>,
+    FC: FCircuit<CF1<C1>>,
+> {
+    pub _gc2: PhantomData<GC2>,
+    pub poseidon_config: PoseidonConfig<CF1<C1>>,
+    pub pp_hash: CF1<C1>,
+    pub i: CF1<C1>,
+    pub i_usize: usize,
+    pub z_0: Vec<CF1<C1>>,
+    pub z_i: Vec<CF1<C1>>,
+    pub external_inputs: Vec<CF1<C1>>,
+    pub F: FC, // F circuit
+    pub u_i_phi: C1,
+    pub U_i: CommittedInstance<C1>,
+    pub U_i1_phi: C1,
+    pub F_coeffs: Vec<CF1<C1>>,
+    pub K_coeffs: Vec<CF1<C1>>,
+    pub x: Option<CF1<C1>>, // public input (u_{i+1}.x[0])
+
+    pub phi_stars: Vec<C1>,
+
+    pub cf1_u_i_cmW: C2,                        // input
+    pub cf2_u_i_cmW: C2,                        // input
+    pub cf_U_i: CycleFoldCommittedInstance<C2>, // input
+    pub cf1_cmT: C2,
+    pub cf2_cmT: C2,
+    pub cf_x: Option<CF1<C1>>, // public input (u_{i+1}.x[1])
```

Comment:
Fixed in https://github.com/privacy-scaling-explorations/sonobe/commit/5483ad01867d06f49f5df9251aaa76d5ac7e864a 😄

Reply:
Honestly I think that having them as `pub` (not `pub(crate)`) was already good. Because anyways in some time somebody is going to open a PR to make them public again so that they can use them to do custom stuff. 
t the end this is a project to allow people to experiment (and even in a non-experimental project, people end up needing to access to internal params for things that intial developments could not predict).
But I don't want to bother you more @winderica , so I'm happy with what you prefer.